pkg/templates/nstemplatetiers/nstemplatetier_generator.go

```diff
@@ -23,74 +24,110 @@ import (
+	if err != nil {
+		return errors.Wrap(err, "unable to create TierTemplates")
 	}
-	// create the NSTemplateTiers
-	nstmplTiersByTier := newNSTemplateTiers(namespace, tierTmplsByTier)
-	cl := commonclient.NewApplyClient(client, s)
 
-	for _, nstmplTier := range nstmplTiersByTier {
-		createdOrUpdated, err := cl.CreateOrUpdateObject(nstmplTier, true, nil)
-		if err != nil {
-			return errors.Wrapf(err, "unable to create or update the '%s' NSTemplateTiers in namespace '%s'", nstmplTier.Name, nstmplTier.Namespace)
-		}
-		if createdOrUpdated {
-			log.Info("NSTemplateTier resource created/updated", "namespace", nstmplTier.Namespace, "name", nstmplTier.Name)
-		} else {
-			log.Info("NSTemplateTier resource was already up-to-date", "namespace", nstmplTier.Namespace, "name", nstmplTier.Name, "ResourceVersion", nstmplTier.ResourceVersion)
-		}
+	// create the NSTemplateTier resources
+	err = generator.createNSTemplateTiers()
+	if err != nil {
+		return errors.Wrap(err, "unable to create NSTemplateTiers")
 	}
+
 	return nil
 }
 
-// templates: namespaces and other cluster-scoped resources belonging to a given tier ("advanced", "basic", "team", etc.)
+type tierGenerator struct {
+	client          client.Client
+	namespace       string
+	scheme          *runtime.Scheme
+	templatesByTier map[string]*tierData
+}
+
+type tierData struct {
+	rawTemplates   *templates
+	tierTemplates  []*toolchainv1alpha1.TierTemplate
+	nstmplTierObjs []commonclient.ToolchainObject
+}
+
+// templates: namespaces and other cluster-scoped resources belonging to a given tier ("advanced", "basic", "team", etc.) and the NSTemplateTier that combines them
 type templates struct {
 	namespaceTemplates map[string]template // namespace templates (including roles, etc.) indexed by type ("dev", "code", "stage")
 	clusterTemplate    *template           // other cluster-scoped resources, in a single template file
+	nsTemplateTier     *template           // NSTemplateTier resource with tier-scoped configuration and references to namespace and cluster templates in its spec, in a single template file
 }
 
-// template: a template content and its latest git revision
+// template: a template's content and its latest git revision
 type template struct {
 	revision string
 	content  []byte
 }
 
-// loadAssetsByTiers loads the assets and dispatches them by tiers, assuming the given `assets` has the following structure:
+// newTierGenerator loads templates from the provided assets and processes the tierTemplates and NSTemplateTiers
+func newTierGenerator(s *runtime.Scheme, client client.Client, namespace string, assets assets.Assets) (*tierGenerator, error) {
+	// load templates from assets
+	templatesByTier, err := loadTemplatesByTiers(assets)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &tierGenerator{
+		client:          client,
+		namespace:       namespace,
+		scheme:          s,
+		templatesByTier: templatesByTier,
+	}
+
+	// process tierTemplates
+	err = c.newTierTemplates()
```

Comment:
IMO it's a bit confusing to have these calls to `c.newTierTemplates()` and `c.newNSTemplateTiers()` here in the `newTierGenerator(...)` func, and then calls to `generator.createTierTemplates()` and `generator.createNSTemplateTiers()` in the `CreateOrUpdateResources` func above.
Would it be possible to merge the `c.newTierTemplates()` with `generator.createTierTemplates()` funcs and `c.newNSTemplateTiers()` with `generator.createNSTemplateTiers()`? 

Or maybe more simply, have `generator.createTierTemplates()` call `c.newTierTemplates()` and `generator.createNSTemplateTiers()` call `c.newNSTemplateTiers()` ?

Reply:
The reason I call `c.newTierTemplates()` and `c.newNSTemplateTiers()` in the `newTierGenerator(...)` is so that if there were a problem with one of the templates we process all of the templates and error out before any resources are created on the system. But I can see how the current way is a bit confusing. How about if I merge `c.newTierTemplates()` and `c.newNSTemplateTiers()` into 1 function eg. `initGenerator()` and that is called when `generator.createTierTemplates()` or `generator.createNSTemplateTiers()` are both called?

Reply:
> The reason I call `c.newTierTemplates()` and `c.newNSTemplateTiers()` in the `newTierGenerator(...)` is so that if there were a problem with one of the templates we process all of the templates and error out before any resources are created on the system.

it's a fair point, indeed. But I believe that the resources potentially created before an error occurs will be "recreated" at the next deployment (well, updated with probably no change is more accurate)

> How about if I merge `c.newTierTemplates()` and `c.newNSTemplateTiers()` into 1 function eg. `initGenerator()` and that is called when `generator.createTierTemplates()` or `generator.createNSTemplateTiers()` are both called?

yes, that could be an option, indeed. Or that could remain in the `newTierGenerator(...)` func, but maybe named differently to avoid the confusion with the names?

Reply:
So I tried a few different things and ended up renaming `c.newTierTemplates()` to `c.initTierTemplates()` and `c.newNSTemplateTiers()` to `c.initNSTemplateTiers()` and then creating a new `c.initTemplateData()` function that calls both of them. The `newTierGenerator()` func now calls `c.initTemplateData()`. You can see the change in https://github.com/codeready-toolchain/host-operator/pull/269/commits/01de1ec7c2c308aa3bd0d0647b2b283d01126623
Do you think it's less confusing now?

Reply:
yes, it's cleaner like this indeed, thanks!